Sources/NIO/Codec.swift

```diff
@@ -101,115 +129,266 @@ private extension ChannelHandlerContext {
         }
     }
+}
 
-    /// Call `decodeLast` before forward the event through the pipeline.
-    public func channelInactive(ctx: ChannelHandlerContext) {
-        if self.cumulationBuffer != nil {
-            ctx.withThrowingToFireErrorAndClose {
-                try self.decodeLoop(ctx: ctx, decodeFunc: self.decodeLast)
-            }
-            // Once the Channel goes inactive we can just drop all previous buffered data.
-            self.cumulationBuffer = nil
-        }
+public class ByteToMessageHandler<Decoder: ByteToMessageDecoder> {
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = Decoder.InboundOut
 
-        ctx.fireChannelInactive()
+    private enum DecodeMode {
+        /// This is a usual decode, ie. not the last chunk
+        case normal
+
+        /// Last chunk
+        case last
     }
 
-    public func handlerAdded(ctx: ChannelHandlerContext) {
-        self.decoderAdded(ctx: ctx)
+    private enum State {
+        case active
+        case leftoversNeedProcessing
+        case done
     }
 
-    public func handlerRemoved(ctx: ChannelHandlerContext) {
-        if let buffer = self.cumulationBuffer as? InboundOut {
-            ctx.fireChannelRead(self.wrapInboundOut(buffer))
-        } else {
-            /* please note that we're dropping the partially received bytes (if any) on the floor here as we can't
-               send a full message to the next handler. */
+    internal private(set) var decoder: Decoder?
+    private var state: State = .active
+    private var buffer: B2MDBuffer = B2MDBuffer()
+
+    public init(_ decoder: Decoder) {
+        self.decoder = decoder
+    }
+
+    deinit {
+        assert(self.state == .done)
+    }
+}
+
+// MARK: ByteToMessageHandler: Test Helpers
+extension ByteToMessageHandler {
+    internal var cumulationBuffer: ByteBuffer? {
+        return self.buffer._testOnlyOneBuffer()
+    }
+}
+
+// MARK: ByteToMessageHandler's Main API
+extension ByteToMessageHandler {
+    private func withNextBuffer(extraData: ByteBuffer?, _ body: (inout Decoder, inout ByteBuffer) throws -> DecodingState) rethrows -> B2MDBuffer.BufferProcessingResult {
+        if let extraData = extraData {
+            self.buffer.append(buffer: extraData)
+        }
+        switch self.buffer.startProcessing() {
+        case .bufferAlreadyBeingProcessed:
+            return .cannotProcessReentrantly
+        case .nothingAvailable:
+            return .didProcess(.needMoreData)
+        case .available(var buffer):
+            var decoder: Decoder? = nil
+            swap(&decoder, &self.decoder)
+            defer {
+                swap(&decoder, &self.decoder)
+                if buffer.readableBytes > 0 {
+                    if self.decoder!.shouldReclaimBytes(buffer: buffer) {
+                        buffer.discardReadBytes()
+                    }
+                }
+                self.buffer.finishProcessing(remainder: buffer)
+            }
```

Comment:
This `defer` block is very, very large. Are we sure it's clearer than just saving off the return value from `body`?

Reply:
@Lukasa `body` throws so taking this out of the `defer` would it super easy to introduce bugs...